posteriors/laplace/dense_hessian.py

```diff
@@ -0,0 +1,144 @@
+from typing import Any, NamedTuple
+from functools import partial
+import torch
+from optree import tree_map
+from optree.integration.torch import tree_ravel
+
+from posteriors.types import TensorTree, Transform, LogProbFn
+from posteriors.tree_utils import tree_size
+from posteriors.utils import (
+    is_scalar,
+    CatchAuxError,
+)
+
+from torch.func import jacrev, jacfwd
+
+
+def build(
+    log_posterior: LogProbFn,
+    init_prec: torch.Tensor | float = 0.0,
+) -> Transform:
+    """Builds a transform for dense Hessian Laplace.
+
+    Args:
+        log_posterior: Function that takes parameters and input batch and
+            returns the log posterior value (which can be unnormalised)
+            as well as auxiliary information, e.g. from the model call.
+        init_prec: Initial precision matrix.
+            If it is a float, it is defined as an identity matrix
+            scaled by that float.
+
+    Returns:
+        Hessian Laplace transform instance.
+    """
+    init_fn = partial(init, init_prec=init_prec)
+    update_fn = partial(update, log_posterior=log_posterior)
+    return Transform(init_fn, update_fn)
+
+
+class DenseLaplaceState(NamedTuple):
+    """State encoding a Normal distribution over parameters,
+    with a dense precision matrix
+
+    Attributes:
+        params: Mean of the Normal distribution.
+        prec: Precision matrix of the Normal distribution.
+        aux: Auxiliary information from the log_posterior call.
+    """
+
+    params: TensorTree
+    prec: torch.Tensor
+    aux: Any = None
+
+
+def init(
+    params: TensorTree,
+    init_prec: torch.Tensor | float = 0.0,
+) -> DenseLaplaceState:
+    """Initialise Normal distribution over parameters
+    with a dense precision matrix.
+
+    Args:
+        params: Mean of the Normal distribution.
+        init_prec: Initial precision matrix.
+            If it is a float, it is defined as an identity matrix
+            scaled by that float.
+
+    Returns:
+        Initial DenseLaplaceState.
+    """
+
+    if is_scalar(init_prec):
+        num_params = tree_size(params)
+        init_prec = init_prec * torch.eye(num_params, requires_grad=False)
+
+    return DenseLaplaceState(params, init_prec)
+
+
+def update(
+    state: DenseLaplaceState,
+    batch: Any,
+    log_posterior: LogProbFn,
+    epsilon: float = 0.0,
+    rescale: float = 1.0,
+    inplace: bool = False,
+) -> DenseLaplaceState:
+    """Adds the Hessian of the negative log-posterior over given batch.
```

Comment:
I think it would be nice to add a warning to the docstring (here and in `build` too) that the Hessian is not guaranteed to be positive definite and so `epsilon>0` might want to be considered

Reply:
Added!